Fix #5838: Use --target_pattern_file for Bazel commands to avoid comm…

divyansha12 opened the pull request:
Fix #5838
This PR implements the feature request to improve the sharded test and coverage workflows by switching from passing large lists of targets as environment variables to using Bazel's --target_pattern_file flag.

The change:

Generates a file containing the list of test targets instead of concatenating them on the command line or via environment variables.

Updates the Bazel commands in GitHub workflows to use --target_pattern_file=<file> for better readability and maintainability.

Optionally uploads the target pattern files as artifacts for easier debugging and analysis of failed CI shards.

Addresses issues with long command lines and target lists, making the CI workflow more robust and easier to understand.

This aligns with Bazel best practices and addresses the concerns mentioned in issue #5838.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

divyansha12 committed:
Fix #5838: Use --target_pattern_file for Bazel commands to avoid command line length issues

adhiamboperes commented:
Hi @divyansha12, are you ready for a review on this PR? If so, please assign @manas-yu.

divyansha12 commented:
@manas-yu ,@adhiamboperes, please review this PR and let me know if any changes are needed.

manas-yu commented on .github/workflows/code_coverage.yml:
```diff
@@ -323,3 +331,4 @@ jobs:
       - name: Check that coverage status is passed
         if: ${{ needs.compute_changed_files.outputs.can_skip_files != 'true' && needs.evaluate_code_coverage_reports.result != 'success' }}
         run: exit 1
+  
```
add new line at the end of files

manas-yu commented on .github/workflows/code_coverage.yml:
```diff
@@ -298,7 +306,7 @@ jobs:
 
       - name: Upload Generated Markdown Report
         uses: actions/upload-artifact@v4
-        if: ${{ !cancelled() }} # IMPORTANT: Upload reports regardless of success or failure status
```
Revert removal of comment.

manas-yu commented on .github/workflows/code_coverage.yml:
```diff
@@ -247,9 +248,16 @@ jobs:
       - name: Upload Coverage Report Artifact
         uses: actions/upload-artifact@v4
         with:
-          name: coverage-report-${{ env.SHARD_NAME }} # Saving with unique names to avoid conflict
```
Revert removal of comment.

manas-yu commented on .github/workflows/code_coverage.yml:
```diff
@@ -166,6 +166,9 @@ jobs:
           echo "FILE_CACHING_BUCKET=$FILE_CATEGORY" >> $GITHUB_ENV
           echo "CHANGED_FILES=$CHANGED_FILES" >> $GITHUB_ENV
           echo "BAZEL_TEST_TARGETS=$BAZEL_TEST_TARGETS" >> $GITHUB_ENV
```
after switching to `--target_pattern_file=bazel_targets.txt` nothing reads the `BAZEL_TEST_TARGETS` env var anymore, we can drop the `echo "BAZEL_TEST_TARGETS=..." >> $GITHUB_ENV` lines.

manas-yu commented on .github/workflows/code_coverage.yml:
```diff
@@ -166,6 +166,9 @@ jobs:
           echo "BAZEL_TEST_TARGETS=$BAZEL_TEST_TARGETS" >> $GITHUB_ENV
+          # Write targets to file for --target_pattern_file
+          echo "$BAZEL_TEST_TARGETS" > bazel_targets.txt
+          echo "Target file created with $(wc -l < bazel_targets.txt) targets"
```
Just keep `echo "$BAZEL_TEST_TARGETS" > bazel_targets.txt`

manas-yu reviewed (changes requested):
Hi @divyansha12 thanks for the changes, PTAL at the review comments.

divyansha12 committed:
Address review comments: remove unused env vars and restore comments

divyansha12 commented:
Hi @manas-yu Thank you for the review! I've addressed all the comments:

Removed the `echo "BAZEL_TEST_TARGETS=..." >> $GITHUB_ENV` lines from both workflows since nothing reads this env var after switching to `--target_pattern_file`
 Removed the target count echo lines (`echo "Target file created with $(wc -l < bazel_targets.txt) targets"`)
 Restored the important comments that were previously removed
Added newlines at the end of both files
All changes have been pushed. Please review when you get a chance. Thanks.

manas-yu commented on .github/workflows/code_coverage.yml:
```diff
@@ -322,4 +328,4 @@ jobs:
       - name: Check that coverage status is passed
         if: ${{ needs.compute_changed_files.outputs.can_skip_files != 'true' && needs.evaluate_code_coverage_reports.result != 'success' }}
-        run: exit 1
+        run: exit 1
```
Add new line at file end.

divyansha12 replied:
Done.

manas-yu commented on .github/workflows/code_coverage.yml:
```diff
@@ -165,7 +165,8 @@ jobs:
           echo "CHANGED_FILES=$CHANGED_FILES" >> $GITHUB_ENV
-          echo "BAZEL_TEST_TARGETS=$BAZEL_TEST_TARGETS" >> $GITHUB_ENV
+          # Write targets to file for --target_pattern_file
+          echo "$BAZEL_TEST_TARGETS" > bazel_targets.txt
```
write targets one per line to bazel_targets.txt (e.g., `printf '%s\n' $BAZEL_TEST_TARGETS > bazel_targets.txt`) so Bazel can parse them correctly.

divyansha12 replied:
Done.

manas-yu commented on .github/workflows/unit_tests.yml:
```diff
@@ -242,6 +239,13 @@ jobs:
+        if: always()
+        uses: actions/upload-artifact@v4
+        with:
+          name: unit-tests-bazel-targets-${{ matrix.affected-tests-bucket-base64-encoded-shard }}
```
Derive a safe SHARD_NAME from just the shard prefix before ; and use that in the artifact name. The artifact name for the uploaded Bazel target file includes the full affected-tests-bucket-base64-encoded-shard value, whose base64 portion contains /, which GitHub doesn’t allow in artifact names

divyansha12 replied:
Done.

manas-yu reviewed (changes requested):
Hi @divyansha12, just one thing to keep in mind for the future don't resolve comments instead address them with a comment, like "done" or "fixed", just a thing we follow here.

also, PTAL at the review comments.

divyansha12 committed:
Fix review comments: use printf for one-target-per-line formatting and add newline at file end

divyansha12 committed:
Add newline at end of unit_tests.yml

divyansha12 commented:
Hi @manas-yu I have addressed all your review comments in commit 34c06d6. Could you please review again when you get a chance? Thanks!

manas-yu commented:
Thanks @divyansha12 changes looks good to me. @adhiamboperes PTAL.

Subham-KRLX commented:
@manas-yu @adhiamboperes  please merge this pr i think its ready to merge.

adhiamboperes reviewed (commented):
Thanks @divyansha12 for this!

I have a question -- since both workflows write to the same file, do they duplicate the targets, or does the code coverage workflow overwrite the contents added by the unit tests workflow? Can you please share a link to the uploaded artifact?

divyansha12 commented:
@adhiamboperes Thanks for the question.

Since GitHub Actions jobs run in isolated environments (separate virtual machines or containers) the unit_tests and code_coverage workflows operate on completely independent file systems. The bazel_targets.txt file created in one workflow does not exist in the other so there is no risk of them overwriting each other or duplicating targets.

adhiamboperes commented:
Thanks @divyansha12. Sounds good to me.

adhiamboperes reviewed (approved):
LGTM, thanks!

Will merge this later tonight.

Please feel free to begin working on another issue or reach out to @manas-yu for guidance.

adhiamboperes commented:
@divyansha12, in future, please do not change the PR template. The essential checklist is a mandatory part of the PR description. In this instance, I have added it back for you.

manas-yu commented:
Hi @divyansha12, congratulations on your first PR! 🎉
It would be great to have you on the Dev-Workflow team at Oppia. Please fill out the [onboarding form](https://docs.google.com/forms/d/e/1FAIpQLSes4wj7oKtiN0iyUwd-Xt_DAOr1b-MHG9_YYIavWBF4G6g8uA/viewform) so we can collect your details and add you to the relevant groups.
